webapp-django/crashstats/topcrashers/views.py

```diff
@@ -0,0 +1,302 @@
+import datetime
+import isodate
+from collections import defaultdict
+
+from django import http
+from django.conf import settings
+from django.shortcuts import render, redirect
+from django.utils import timezone
+from django.utils.http import urlquote
+
+from waffle.decorators import waffle_switch
+
+from session_csrf import anonymous_csrf
+
+from crashstats.crashstats import models
+from crashstats.crashstats.decorators import (
+    check_days_parameter,
+    pass_default_context,
+)
+from crashstats.supersearch.form_fields import split_on_operator
+from crashstats.supersearch.models import (
+    SuperSearchUnredacted,
+)
+
+
+def get_date_boundaries(parameters):
+    """Return the date boundaries in a set of parameters.
+
+    Return a tuple with 2 datetime objects, the first one is the lower bound
+    date and the second one is the upper bound date.
+    """
+    default_date_range = datetime.timedelta(days=7)
+
+    greater_than = None
+    lower_than = None
+
+    if not parameters.get('date'):
+        lower_than = timezone.now()
+        greater_than = lower_than - default_date_range
+    else:
+        for param in parameters['date']:
+            value = isodate.parse_datetime(split_on_operator(param)[1])
+
+            if (
+                '<' in param and (
+                    not lower_than or
+                    (lower_than and lower_than > value)
+                )
+            ):
+                lower_than = value
+            if (
+                '>' in param and (
+                    not greater_than or
+                    (greater_than and greater_than < value)
+                )
+            ):
+                greater_than = value
+
+        if not lower_than:
+            # add a lower than that is now
+            lower_than = timezone.now()
+
+        if not greater_than:
+            # add a greater than that is lower_than minus the date range
+            greater_than = lower_than - default_date_range
+
+    return (greater_than, lower_than)
+
+
+def get_topcrashers_results(**kwargs):
+    '''Return the results of a search. '''
+    results = []
+
+    params = kwargs
+    params['_aggs.signature'] = [
+        'platform',
+        'is_garbage_collecting',
+        'hang_type',
+        'process_type',
+    ]
```

Comment:
The magic sauce! I'll try to remember this. But probably will forget and come begging for your help later.